internal/core/command/device.go

```diff
@@ -63,70 +74,97 @@ func executeCommandByDeviceID(
+// successfully extracted, the associated error is returned.
+func extractDeviceIdAndCommandIdFromRequest(r *http.Request) (string, string, error) {
+	vars := mux.Vars(r)
+	deviceID := vars[ID]
+	commandID := vars[COMMANDID]
+
+	if deviceID == "" || commandID == "" {
+		return deviceID, commandID, errors.NewErrExtractingInfoFromRequest()
 	}
 
-	return executeCommandByDevice(ctx, d, c, body, queryParams, isPutCommand, lc)
+	return deviceID, commandID, nil
 }
 
 func executeCommandByName(
+	originalRequest *http.Request,
 	ctx context.Context,
 	dn string,
 	cn string,
 	body string,
-	queryParams string,
-	isPutCommand bool,
 	lc logger.LoggingClient,
 	dbClient interfaces.DBClient,
-	deviceClient metadata.DeviceClient) (string, error) {
+	deviceClient metadata.DeviceClient,
+	httpCaller internal.HttpCaller) (deviceServiceResponse *http.Response, failure error) {
 
 	d, err := deviceClient.DeviceForName(ctx, dn)
 	if err != nil {
-		return "", err
+		return nil, err
 	}
 
 	if d.AdminState == contract.Locked {
-		return "", errors.NewErrDeviceLocked(d.Name)
+		return nil, errors.NewErrDeviceLocked(d.Name)
 	}
 
 	command, err := dbClient.GetCommandByNameAndDeviceId(cn, d.Id)
 	if err != nil {
-		return "", err
+		return nil, err
 	}
 
-	return executeCommandByDevice(ctx, d, command, body, queryParams, isPutCommand, lc)
+	return executeCommandByDevice(ctx, d, command, body, lc, originalRequest, httpCaller)
 }
 
 func executeCommandByDevice(
 	ctx context.Context,
 	device contract.Device,
 	command contract.Command,
 	body string,
-	queryParams string,
-	isPutCommand bool,
-	lc logger.LoggingClient) (string, error) {
+	lc logger.LoggingClient,
+	originalRequest *http.Request,
+	httpCaller internal.HttpCaller) (deviceServiceResponse *http.Response, failure error) {
 
+	var method string
 	var ex Executor
 	var err error
-	if isPutCommand {
-		ex, err = NewPutCommand(device, command, body, ctx, &http.Client{}, lc)
-	} else {
-		ex, err = NewGetCommand(device, command, queryParams, ctx, &http.Client{}, lc)
+
+	if originalRequest == nil {
+		return nil, errors.NewErrParsingOriginalRequest("method")
+	}
+
+	switch originalRequest.Method {
+	case http.MethodPut:
+		ex, err = NewPutCommand(device, command, body, ctx, httpCaller, lc, originalRequest)
+	case http.MethodGet:
+		ex, err = NewGetCommand(device, command, ctx, httpCaller, lc, originalRequest)
+	default:
+		lc.Error(fmt.Sprintf("unknown method: %s", method))
 	}
 
 	if err != nil {
-		return "", err
+		return nil, err
 	}
 
-	responseBody, responseCode, err := ex.Execute()
+	deviceServiceResponse, err = ex.Execute()
 	if err != nil {
-		return "", err
+		return nil, err
 	}
-	if responseCode != http.StatusOK {
-		return "", types.NewErrServiceClient(responseCode, []byte(responseBody))
+
+	buf := new(bytes.Buffer)
+	_, readErr := buf.ReadFrom(deviceServiceResponse.Body)
+	if readErr != nil {
+		return nil, readErr
```

Comment:
I think you have a bug here. You are reading the response and not returning it and causing the caller to read the body from the response later, which will fail because you can only read it one. This is backed with the fact that these changes do not work. Issuing a GET command to device virtual for binary data though core-command results in an empty response body with a 200 status code, with the proper header. This was not noticed at first due to the generic name of `buf`. Can we please name variables properly so we can spot bugs a little easier?

Reply:
Absolutely, I'm on it 👍